File: Theme 7/1.cpp
#include <stdio.h>
#include <math.h>
#include <stdlib.h>

double eval1(double x, double y)
{
    return (-1 * y * tan(x) + sin(x * 2));
}

double *Euler_Cauchy(double a, double b, double h, double y0, double (*eval)(double, double))
{
    int n = abs((b - a) / h);
    double *y = (double *)malloc(sizeof(double) * n + 1);
    double x = a;
    y[0] = y0;
    for (int i = 0; i < n; i++)
    {
        y[i+1] = (y[i] + h * eval(x, y[i]));
        x += h;
    }
    return y;
}

double prec_eval1(double x)
{
    return -2 * pow(cos(x), 2) + cos(x);
}

void print_answer(double a, double b, double h, double *arr, double (*eval)(double))
{
    double x = a;
    int n = abs((b - a) / h);
    for (int i = 0; i < n+1; i++)
    {
        printf("Вычисленный ответ в точке х = %lf: %lf Точный: %lf Разница: %lf\n", x, arr[i], eval(x), eval(x) - arr[i]);
        x += h;
    }
}

int main()
{
    double *arr = Euler_Cauchy(0, 1, 0.1, -1, eval1);
    print_answer(0, 1, 0.1, arr, prec_eval1);
    free(arr);
}

File: Theme 7/1.py
import math


def eval1(x, y):
    return -1 * y * math.tan(x) + math.sin(x * 2)


def Euler_Cauchy(a, b, h, y0, eval):
    y = [y0]
    n = int(abs(b - a) / h)
    x = a
    for i in range(n):
        y.append(y[i] + h * eval(x, y[i]))
        x += h
    return y


def prec_eval1(x):
    return -2 * math.cos(x) ** 2 + math.cos(x)


def print_answer(a, h, arr, eval):
    x = a
    for i in arr:
        print(f"Вычисленный ответ в точке х = {x:.1f}: {i} Точный: {eval(x)} Разница: {eval(x)-i}")
        x += h


arr = Euler_Cauchy(0, 1, 0.1, -1, eval1)
print_answer(0, 0.1, arr, prec_eval1)
